mutation.hh

```diff
@@ -276,20 +265,34 @@ std::optional<stop_iteration> consume_clustering_fragments(schema_ptr s, mutatio
             ++crs_it;
             continue;
         }
-        bool emit_rt;
-        if (crs_it != crs_end && rts_it != rts_end) {
-            const auto cmp_res = cmp(rts_it->position(), crs_it->position());
-            if constexpr (reverse == consume_in_reverse::legacy_half_reverse) {
-                emit_rt = cmp_res > 0;
+        range_tombstone* rt_to_emit = nullptr;
+        if (rts_it != rts_end) {
+            if constexpr (reverse == consume_in_reverse::yes) {
+                if (!cookie.reversed_rt) {
+                    cookie.reversed_rt = std::move(rts_it->tombstone());
+                    cookie.reversed_rt->reverse();
+                }
+                rt_to_emit = &*cookie.reversed_rt;
             } else {
-                emit_rt = cmp_res < 0;
+                rt_to_emit = &rts_it->tombstone();
+            }
+            if (crs_it != crs_end) {
+                auto cmp_res = cmp(rt_to_emit->position(), crs_it->position());
+                if constexpr (reverse == consume_in_reverse::legacy_half_reverse) {
+                    if (cmp_res < 0) {
+                        rt_to_emit = nullptr;
+                    }
+                } else if (cmp_res > 0) {
+                    rt_to_emit = nullptr;
+                }
```

Comment:
I don't understand the purpose of these changes. I find the new code much harder to follow.

Reply:
I didn't want to reverse the candidate range tombstone in place in full-reverse mode,
hence rt_to_emit was introduced, which either points to the tombstone in the list in other modes or to a reversed one on the stack, in full reverse mode.  With that, we don't need `bool emit_rt`, since the pointer can be used for the same purpose.

Reply:
I'm not sure it was worth it. Before we had two simple if-else statements, one deciding what to emit, and the other one executing it. Now we have 3 levels of ifs some levels having multiple ifs in them.
You didn't need to reverse in place anyway, you could reverse just before emitting.

Reply:
> You didn't need to reverse in place anyway, you could reverse just before emitting.

The range tombstone needs to be reversed also before comparing, can't just compare end_position, because its bound weight needs to be reversed too.

Reply:
Ah. You can reverse in-place BTW. The consumption is destructive, you can mutate the mutation as you like.

Reply:
We are not required/supposed to leave the mutation in a valid state after `consume()` was called. It just has to be in a such a state that we can resume the consume later (if the user wants it).

Reply:
Change reduced in 935bb643ae to minimum needed.